### DISCUSSION TITLE
is there any way to add Document level Javascript?

### DISCUSSION THREAD

**Discussion by deven740** [pymupdf/PyMuPDF#2157] [Looking for help] [answered]
Hi everyone!

 In Adobe, Document level Javascript can be added to the PDF file, can this Javascript be added to the PDF file using PyMuPDF

`var btn = this.getField("btn")
btn.setAction("MouseEnter","this.getOCGs()[0].state=false")
btn.setAction("MouseExit","this.getOCGs()[0].state=true")
`
I want to implement the following scenario

When user opens the PDF file, Javascript code (Document level Javascript code added by PyMuPDF) should get executed which adds a mouse enter and mouse exit event on all the buttons present in the PDF file to show or hide the OCG layers when these events occur.

Thanks in advance.

**Comment by JorjMcKie** [comment, marked as answer]
You have to set `OpenAction` to a JS action dictionary object.
This is currently not implemented as such in PyMuPDF.
But it should be possible to achieve your goal using PyMuPDF low-level interface calls. Approach could be along the lines:
1. Add an object defining the JS action
2. Access the PDF catalog and add the `OpenAction` key pointing to the JS action dictionary object

Define the JS action object:
```python
xref = doc.get_new_xref()
js = "the javascript text"
objsource = f"<</Type/Action/S/JavaScript/JS({js})>>"
doc.update_object(xref, objsource)
```

Update the catalog:
```python
cat = doc.pdf_catalog()
doc.xref_set_key(cat, "OpenAction", f"{xref} 0 R")
```

Then save. Fingers crossed that everything works ... 😎

**Reply by deven740** [reply]
Everything worked perfectly!!

Thanks for your help.
